histomicstk/saliency/cellularity_detection_thresholding.py

```diff
@@ -0,0 +1,683 @@
+            'hsi_thresholds': {
+                'whitespace': {
+                    'hue': {'min': 0, 'max': 1.0},
+                    'saturation': {'min': 0, 'max': 0.2},
+                    'intensity': {'min': 220, 'max': 255},
+                },
+            },
+            'lab_thresholds': {
+                'blue_sharpie': {
+                    'l': {'min': -1000, 'max': 1000},
+                    'a': {'min': -1000, 'max': 1000},
+                    'b': {'min': -1000, 'max': -0.02},
+                },
+                'blood': {
+                    'l': {'min': -1000, 'max': 1000},
+                    'a': {'min': 0.02, 'max': 1000},
+                    'b': {'min': -1000, 'max': 1000},
+                },
+            },
+
+            # for stain unmixing to deconvolove and/or color normalize
+            'stain_unmixing_routine_params': {
+                'stains': ['hematoxylin', 'eosin'],
+                'stain_unmixing_method': 'macenko_pca',
+            },
+
+            # params for getting cellular regions
+            'cellular_step1_sigma': 0.,
+            'cellular_step1_min_size': 100,
+            'cellular_step2_sigma': 1.5,
+            'cellular_largest_n': 5,
+            'cellular_top_n': 2,
+
+            # visualization params
+            'visualize': True,
+            'opacity': 0,
+            'lineWidth': 3.0,
+        }
+        default_attr.update(kwargs)
+        super(Cellularity_detector_thresholding, self).__init__(
+            default_attr=default_attr)
+
+        self.color_normalization_method = \
+            self.color_normalization_method.lower()
+        assert self.color_normalization_method in [
+            'reinhard', 'macenko_pca', 'none']
+
+        # set attribs
+        self.gc = gc
+        self.slide_id = slide_id
+        self.GTcodes = GTcodes
+
+        self.fix_GTcodes()
+
+    # %% ======================================================================
+
+    def fix_GTcodes(self):
+        """Fix self.GTcodes (important!)."""
+        # validate
+        self.GTcodes.index = self.GTcodes.loc[:, "group"]
+        necessary_indexes = self.keep_components + [
+            'outside_tissue', 'not_specified',
+            'maybe_cellular', 'top_cellular']
+        assert all(j in list(self.GTcodes.index) for j in necessary_indexes)
+
+        # Make sure the first things layed out are the "background" components
+        min_val = np.min(self.GTcodes.loc[:, 'overlay_order'])
+        self.GTcodes.loc['outside_tissue', 'overlay_order'] = min_val - 2
+        self.GTcodes.loc['not_specified', 'overlay_order'] = min_val - 1
+
+        # reorder in overlay order (important)
+        self.GTcodes.sort_values('overlay_order', axis=0, inplace=True)
+        self.ordered_components = list(self.GTcodes.loc[:, "group"])
+
+        # only keep relevant components (for HSI/LAB thresholding)
+        for c in self.ordered_components.copy():
```

Comment:
If we want to keep Python 2 compatibility, this should be `[:]` rather than `.copy()`, since it is a Python list.